kotlin/workflow-core/src/main/java/com/squareup/workflow/Worker.kt

```diff
@@ -0,0 +1,239 @@
+ * should be allowed to continue working. [Worker]s that are not included in a render pass are
+ * cancelled.
+ *
+ * ## Example: Network request
+ *
+ * Let's say you have a network service with an API that returns a [Deferred], and you want to
+ * call that service from a [Workflow].
+ *
+ * ```
+ * interface TimeService {
+ *   fun getTime(timezone: String): Deferred<Long>
+ * }
+ * ```
+ *
+ * The first step is to define a [Worker] that can call this service.
+ * ```
+ * class TimeWorker(
+ *   val timezone: String,
+ *   val service: TimeService
+ * ): Worker<Long> {
+ *
+ *   override suspend fun performWork(emitter: Emitter<Long>) {
+ *     val timeDeferred = service.getTime(timezone)
+ *     emitter.emitOutput(timeDeferred.await())
+ *   }
+ * }
+ * ```
+ *
+ * You also need to define how to determine if a previous [Worker] is already doing the same work.
+ * This will ensure that if the same request is made by the same [Workflow] in adjacent render
+ * passes, we'll keep the request alive from the first pass.
+ * ```
+ *   override fun doesSameWorkAs(otherWorker: Worker<*>): Boolean =
+ *     otherWorker is TimeWorker &&
+ *       timezone == otherWorker.timezone
+ * ```
+ *
+ * Now you can request the time from your [Workflow]:
+ * ```
+ * class MyWorkflow(private val timeWorker: TimeWorker) {
+ *   override fun render(…): Foo {
+ *     context.onWorkerOutput(timeWorker) { time -> emitOutput("The time is $time") }
+ *   }
+ * ```
+ *
+ * @see create
+ * @see from
+ * @see fromNullable
+ * @see fromChannel
+ * @see Deferred.asWorker
+ * @see BroadcastChannel.asWorker
+ */
+interface Worker<out T> {
+
+  /**
+   * Passed into [performWork] to allow the [Worker] to emit outputs.
+   */
+  interface Emitter<in T> {
+    /**
+     * Delivers an output value to the [Workflow] that is running this [Worker].
+     */
+    suspend fun emitOutput(output: T)
+  }
+
+  /**
+   * Used by [RenderContext.onWorkerOutputOrFinished] to distinguish between the two events of a
+   * [Worker] emitting an output, and finishing (returning from [performWork]).
+   */
+  sealed class OutputOrFinished<out T> {
+    data class Output<out T>(val value: T) : OutputOrFinished<T>()
+    object Finished : OutputOrFinished<Nothing>()
+  }
+
+  /**
+   * Override this method to do the actual work. This is a suspend function and is invoked in the
+   * context of the workflow runtime. When this [Worker], it's parent [Workflow], or any ancestor
+   * [Workflow]s are torn down, the coroutine in which this function is invoked will be cancelled.
+   * If you need to do cleanup when the [Worker] is cancelled, wrap your work in a `try { }` block
+   * and do your cleanup in a `finally { }` block. For more information on how coroutine
+   * cancellation works, see [the coroutine guide](https://kotlinlang.org/docs/reference/coroutines/cancellation-and-timeouts.html).
+   */
+  suspend fun performWork(emitter: Emitter<T>)
```

Comment:
Should this be `suspend fun Emitter<T>.performWork()` instead?

Reply:
No, to me this clearer as is.